cpp/include/cudf/io/detail/data_casting.cuh

```diff
@@ -295,6 +305,470 @@ process_string(in_iterator_t in_begin,
+      // All escaping backslash should be skipped.
+
+      struct state_table {
+        bool state[2];
+      };
+      // using state_table = bool[2]; Try this. and see if compiler errors
+      using BlockScan = cub::BlockScan<state_table, BLOCK_SIZE>;
+      __shared__ typename BlockScan::TempStorage temp_slash;
+      state_table curr{is_within_bounds && c == '\\', false};  // state transition vector.
+      auto composite_op = [](state_table op1, state_table op2) {
+        return state_table{op2.state[op1.state[0]], op2.state[op1.state[1]]};
+      };
+      state_table scanned;
+      // inclusive scan. TODO both inclusive and exclusive available in cub.
+      BlockScan(temp_slash).InclusiveScan(curr, scanned, composite_op);
+      auto is_escaping_backslash = scanned.state[init_state];
+      if (threadIdx.x == BLOCK_SIZE - 1) init_state = is_escaping_backslash;
+      // There is another __syncthreads() at the end of for-loop.
+      skip |= is_escaping_backslash;
+
+      if (!skip) {
+        if (prev_c != '\\') {
+          this_num_out = 1;
+          if (d_chars) write_char = c;
+        } else {
+          if (escaped_char != UNICODE_SEQ) {
+            this_num_out = 1;
+            write_char   = escaped_char;
+          } else {
+            // Unicode
+            // \uXXXX
+            auto hex_val     = parse_unicode_hex(in_begin + char_index + 1);
+            auto hex_low_val = 0;
+            // \uXXXX\uXXXX
+            if ((in_begin + char_index + 4 + 6) < in_end && in_begin[char_index + 1 + 4] == '\\' &&
+                in_begin[char_index + 1 + 5] == 'u') {
+              hex_low_val = parse_unicode_hex(in_begin + char_index + 1 + 6);
+            }
+            if (hex_val >= UTF16_HIGH_SURROGATE_BEGIN && hex_val < UTF16_HIGH_SURROGATE_END &&
+                hex_low_val >= UTF16_LOW_SURROGATE_BEGIN && hex_low_val < UTF16_LOW_SURROGATE_END) {
+              // Compute UTF16-encoded code point
+              uint32_t unicode_code_point = 0x10000 +
+                                            ((hex_val - UTF16_HIGH_SURROGATE_BEGIN) << 10) +
+                                            (hex_low_val - UTF16_LOW_SURROGATE_BEGIN);
+              write_char   = strings::detail::codepoint_to_utf8(unicode_code_point);
+              this_num_out = strings::detail::bytes_in_char_utf8(write_char);
+            } else {
+              if (hex_val >= UTF16_LOW_SURROGATE_BEGIN && hex_val < UTF16_LOW_SURROGATE_END) {
+                skip         = true;
+                this_num_out = 0;
+                write_char   = 0;
+              } else {
+                // if u8
+                write_char   = strings::detail::codepoint_to_utf8(hex_val);
+                this_num_out = strings::detail::bytes_in_char_utf8(write_char);
+              }
+            }
+          }
+        }
+      }  // !skip end.
+      {
+        using BlockScan2 = cub::BlockScan<size_type, BLOCK_SIZE>;
```

Comment:
same as char2, please use a more descriptive name.

Reply:
updated in the merged code.